File: backend/app/schemas/stats.py
# app/schemas/stats.py
from datetime import date
from pydantic import BaseModel


class SummaryStats(BaseModel):
    period: str  # "month" or "year"
    year: int
    month: int | None = None
    total_spent: float
    avg_per_day: float
    transactions_count: int
    top_category: str | None = None
    top_category_amount: float | None = None


class MonthlyPoint(BaseModel):
    month: int  # 1-12
    total_amount: float


class MonthlyStats(BaseModel):
    year: int
    points: list[MonthlyPoint]


class DailyPoint(BaseModel):
    day: int # 1-31
    date: date
    total_amount: float


class DailyStats(BaseModel):
    year: int
    month: int
    points: list[DailyPoint]


class CategoryPoint(BaseModel):
    category_id: int | None
    category_name: str | None
    total_amount: float


class CategoryStats(BaseModel):
    year: int
    month: int | None = None
    categories: list[CategoryPoint]

File: web/src/store/stats.ts
import { create } from 'zustand';
import api from '@/lib/api';

// Types mirror the backend Pydantic schemas
interface SummaryStats {
    period: string;
    year: number;
    month: number | null;
    total_spent: number;
    avg_per_day: number;
    transactions_count: number;
    top_category: string | null;
    top_category_amount: number | null;
}

interface MonthlyPoint {
    month: number;
    total_amount: number;
}

interface MonthlyStats {
    year: number;
    points: MonthlyPoint[];
}

interface CategoryPoint {
    category_id: number | null;
    category_name: string | null;
    total_amount: number;
}

interface CategoryStats {
    year: number;
    month: number | null;
    categories: CategoryPoint[];
}

interface StatsState {
    summary: SummaryStats | null;
    monthly: MonthlyStats | null;
    categories: CategoryStats | null;
    isLoading: boolean;
    error: string | null;

    fetchAllStats: () => Promise<void>;
}

export const useStatsStore = create<StatsState>((set) => ({
    summary: null,
    monthly: null,
    categories: null,
    isLoading: false,
    error: null,

    fetchAllStats: async () => {
        set({ isLoading: true, error: null });
        try {
            const [summaryRes, monthlyRes, categoryRes] = await Promise.all([
                api.get('/stats/summary'),
                api.get('/stats/monthly'),
                api.get('/stats/categories')
            ]);

            set({
                summary: summaryRes.data,
                monthly: monthlyRes.data,
                categories: categoryRes.data,
                isLoading: false
            });
        } catch (error: any) {
            console.error("Stats fetch failed:", error);
            set({
                error: error.message || 'Failed to fetch stats',
                isLoading: false
            });
        }
    }
}));
